art/attacks/evasion/graphite/graphite_blackbox.py

```diff
@@ -0,0 +1,889 @@
+            raise ValueError("focal must be int or float")
+        focal = float(focal)
+
+        # Check the mask
+        if mask is not None:
+            if len(mask.shape) == len(x.shape):
+                mask = mask.astype(ART_NUMPY_DTYPE)
+            else:
+                mask = np.array([mask.astype(ART_NUMPY_DTYPE)] * x.shape[0])
+        else:
+            mask = np.array([None] * x.shape[0])
+
+        # Get clip_min and clip_max from the classifier or infer them from data
+        if self.estimator.clip_values is not None:
+            clip_min, clip_max = self.estimator.clip_values
+        else:
+            clip_min, clip_max = np.min(x), np.max(x)
+
+        # target initialization image
+        x_tar = kwargs.get("x_tar")
+
+        # Some initial setups
+        x_adv = x.astype(ART_NUMPY_DTYPE)
+
+        if self.estimator.channels_first:
+            x = np.transpose(x, (0, 2, 3, 1))
+            x_adv = np.transpose(x_adv, (0, 2, 3, 1))
+            x_tar = np.transpose(x_tar, (0, 2, 3, 1))
+            if len(mask.shape) == 4:
+                mask = np.transpose(mask, (0, 2, 3, 1))
+
+        y = np.argmax(y, axis=1)
+
+        # Generate the adversarial samples
+        for i in range(x_adv.shape[0]):
+            x_adv[i] = self._perturb(
+                x=x_adv[i],
+                y=y[i],
+                x_tar=x_tar[i],
+                mask=mask[i],
+                pts=pts,
+                obj_width=obj_width,
+                focal=focal,
+                clip_min=clip_min,
+                clip_max=clip_max,
+            )
+
+        y = to_categorical(y, self.estimator.nb_classes)  # type: ignore
+
+        # COMPUTE SUCCESS RATE
+        x_copy = np.zeros((x.shape[0], self.noise_size[1], self.noise_size[0], x.shape[3]))
+        x_adv_copy = np.zeros((x_adv.shape[0], self.noise_size[1], self.noise_size[0], x_adv.shape[3]))
+        for i in range(x_copy.shape[0]):
+            x_copy[i] = convert_to_network(x[i], self.net_size, clip_min, clip_max)
+            x_adv_copy[i] = convert_to_network(x_adv[i], self.net_size, clip_min, clip_max)
+
+        if self.estimator.channels_first:
+            x_copy = np.transpose(x_copy, (0, 3, 1, 2))
+            x_adv_copy = np.transpose(x_adv_copy, (0, 3, 1, 2))
+
+        logger.info(
+            "Success rate of GRAPHITE hard-label attack: %.2f%%",
+            100
+            * compute_success(
+                self.estimator,
+                x_copy.astype(np.float32),
+                y,
+                x_adv_copy.astype(np.float32),
+                self.targeted,
+                batch_size=self.batch_size,
+            ),
+        )
+
+        if self.estimator.channels_first:
+            x_adv = np.transpose(x_adv, (0, 3, 1, 2))
+
+        return x_adv
+
+    def _perturb(
+        self,
+        x: np.ndarray,
+        y: int,
+        x_tar: np.ndarray,
+        mask: Optional[np.ndarray],
```

Comment:
I think if you are using optional here, you should explicitly set the default value to None. You might also want to move it to the end of the arg list. This also occurs on lines 311, 373, 502, 547, 593, 656, 731,